config/v1/types_authentication.go

```diff
@@ -271,6 +296,99 @@ type TokenClaimMapping struct {
 	Claim string `json:"claim"`
 }
 
+// +kubebuilder:validation:XValidation:rule="!(has(self.claim) && has(self.expression))",message="both claim and expression must not be set"
+// +kubebuilder:validation:XValidation:rule="has(self.claim) || has(self.expression)",message="either claim or expression must be set"
+type TokenClaimOrExpressionMapping struct {
+	// claim is an optional field for specifying the
+	// JWT token claim that is used in the mapping.
+	// The value of this claim will be assigned to
+	// the field in which this mapping is associated.
+	//
+	// Either claim or expression must be set.
+	// claim must not be specified when expression is set.
+	//
+	// +optional
+	Claim string `json:"claim"`
+
+	// expression is an optional field for specifying a
+	// CEL expression that produces a string value from
+	// JWT token claims.
+	//
+	// CEL expressions have access to the token claims
+	// through a CEL variable, 'claims'.
+	// 'claims' is a map of claim names to claim values.
+	// For example, the 'sub' claim value can be accessed as 'claims.sub'.
+	// Nested claims can be accessed using dot notation ('claims.foo.bar').
+	//
+	// Either claim or expression must be set.
+	// expression must not be specified when claim is set.
+	// When specified, expression must be at least 1 character in length
+	// and must not exceed 1024 characters in length.
+	//
+	// +optional
+	// +kubebuilder:validation:MaxLength=1024
+	// +kubebuilder:validation:MinLength=1
+	Expression string `json:"expression,omitempty"`
```

Comment:
Are there any syntactic validations we could add here, eg, a valid character set?

Is 1024 enough? CEL expressions can be quite verbose, maybe 4096?

Reply:
Is there a known regex for CEL expressions? I went looking but couldn't find anything concrete.

I'm happy to increase to 4096. I chose 1024 arbitrarily.

Reply:
Probably. Currently the kube apiserver will validate that on startup and fail to start if the CEL expression won't compile. We could also validate this at runtime in the hypershift control-plane-operator and the openshift cluster-authentication-operator.

Reply:
It would certainly be preferable to validate the expression before KAS fails to start up IMO, should think about how we might achieve that

Reply:
Agreed. Currently my thinking is to do this at runtime in the controllers until we have a better way to check this at admission time. This will mean we don't attempt to roll out changes to the KAS if it doesn't compile and the associated operator status should convey the error to a user.

Reply:
Here is a commit to my hypershift PR adding compile checks at runtime: https://github.com/openshift/hypershift/pull/5840/commits/dff4229f9fbcaacdfbd39c56460a3f304761b6f9

Reply:
Joel mentioned in slack that admission time is preferable but runtime is OK here. Should use admission time checks via an admission webhook for standalone GA.